src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java

```diff
@@ -302,13 +300,17 @@ protected void retrieve(@CheckForNull SCMSourceCriteria criteria, SCMHeadObserve
     protected SCMRevision retrieve(SCMHead head, TaskListener listener)
             throws IOException, InterruptedException {
         if (head instanceof BitbucketPullRequestSCMHead) {
-            return new BitbucketPullRequestSCMRevision((BitbucketPullRequestSCMHead) head);
+            return fetchBitbucketPullRequest((BitbucketPullRequestSCMHead) head).map(fetchedPullRequest -> {
```

Comment:
Note: if the PR is deleted, and build is run again, we get a 404 here. Console output (same behavior for deleted tags):
```
Started by user unknown or anonymous
com.atlassian.bitbucket.jenkins.internal.client.exception.NotFoundException: - response: 404
	at com.atlassian.bitbucket.jenkins.internal.http.HttpRequestExecutorImpl.handleError(HttpRequestExecutorImpl.java:153)
	at com.atlassian.bitbucket.jenkins.internal.http.HttpRequestExecutorImpl.makeRequest(HttpRequestExecutorImpl.java:108)
	at com.atlassian.bitbucket.jenkins.internal.http.HttpRequestExecutorImpl.performRequest(HttpRequestExecutorImpl.java:173)
	at com.atlassian.bitbucket.jenkins.internal.http.HttpRequestExecutorImpl.executeRequest(HttpRequestExecutorImpl.java:94)
	at com.atlassian.bitbucket.jenkins.internal.http.HttpRequestExecutorImpl.executeGet(HttpRequestExecutorImpl.java:65)
	at com.atlassian.bitbucket.jenkins.internal.client.BitbucketRequestExecutor.makeGetRequest(BitbucketRequestExecutor.java:185)
	at com.atlassian.bitbucket.jenkins.internal.client.BitbucketRequestExecutor.makeGetRequest(BitbucketRequestExecutor.java:107)
	at com.atlassian.bitbucket.jenkins.internal.client.BitbucketTagClientImpl.getRemoteTag(BitbucketTagClientImpl.java:72)
	at com.atlassian.bitbucket.jenkins.internal.scm.BitbucketSCMSource.lambda$fetchBitbucketTag$13(BitbucketSCMSource.java:497)
```

If that's undesired, I'll have to update the get[PullRequest|Commit|Tag] methods to catch the NotFoundException and then return null in that case (or change the Clients to return optional perhaps) e.g. in https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/pull/445/files#diff-ce314f448c8c2164c49084a9b703df223c3c1cf11a72b57bf974b712ef94476fR31

Reply:
Can you explain this a bit more? If the PR is deleted and the build is run again, unless it's done before Jenkins unmarshals the webhook I'd expect the build to be disabled. Did you disable the trigger to do this?

Reply:
> Did you disable the trigger to do this

Yep, all testing was done with no webhook configured

Reply:
Got it, manual trigger. I notice actually, to make this happen not only must you not listen to a webhook, but not _have_ a webhook at all. This is an edge case- if a webhook is configured the job will be disabled so this condition won't happen.
Unfortunately, on a manual trigger we don't observe the head, and I think this is probably because Jenkins assumes you have already (that's what our webhook is for)- for me the ideal behaviour here is we fail to observe the head and disable the job, but I don't see a way to do that. The 404 in the build logs isn't great.
I think wrapping each of these three fetches in a NotFoundException (rather than a BitbucketClientException, I think a lot of those exceptions SHOULD throw a stack trace in the build logs)- and on catch print an error message to the task listener and returning null is probably the best we can do here. The error can prompt the user to scan the multibranch again- which should be unable to observe the missing head and disable the job.

Reply:
That's a good point, the console output after your suggested changes is much clearer:
```
Started by user unknown or anonymous
ERROR: The requested resource does not exist
ERROR: Could not determine exact tip revision of hello; falling back to nondeterministic checkout
...
ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job.
ERROR: Maximum checkout retry attempts reached, aborting
Finished: FAILURE
```